services/sharepoint-connector/src/microsoft-apis/graph/middlewares/metrics.middleware.ts

```diff
@@ -5,66 +5,156 @@ import {
+import type { Counter, Histogram } from '@opentelemetry/api';
+import type { MetricService } from 'nestjs-otel';
+import type { Config } from '../../../config';
+import {
+  createApiMethodExtractor,
+  getDurationBucket,
+  getHttpStatusCodeClass,
+} from '../../../utils/metrics.util';
+import { elapsedMilliseconds, elapsedSeconds } from '../../../utils/timing.util';
 import { GraphApiErrorResponse, isGraphApiError } from '../types/sharepoint.types';
 
 export class MetricsMiddleware implements Middleware {
   private readonly logger = new Logger(this.constructor.name);
   private nextMiddleware: Middleware | undefined;
+  private readonly extractApiMethod: ReturnType<typeof createApiMethodExtractor>;
+
+  private readonly spcGraphApiRequestDurationSeconds: Histogram;
+  private readonly spcGraphApiThrottleEventsTotal: Counter;
+  private readonly spcGraphApiSlowRequestsTotal: Counter;
+
+  private readonly msTenantId: string;
+
+  public constructor(metricService: MetricService, configService: ConfigService<Config, true>) {
+    this.spcGraphApiRequestDurationSeconds = metricService.getHistogram(
+      'spc_graph_api_request_duration_seconds',
+      {
+        description: 'Performance of all Microsoft Graph API calls',
+      },
+    );
+
+    this.spcGraphApiThrottleEventsTotal = metricService.getCounter(
+      'spc_graph_api_throttle_events_total',
+      {
+        description: 'Total number of Graph API throttling events',
+      },
+    );
+
+    this.spcGraphApiSlowRequestsTotal = metricService.getCounter(
+      'spc_graph_api_slow_requests_total',
+      {
+        description: 'Total number of slow Graph API requests',
+      },
+    );
+
+    this.msTenantId = configService.get('sharepoint.authTenantId', { infer: true });
+
+    this.extractApiMethod = createApiMethodExtractor([
+      'sites',
+      'drives',
+      'items',
+      'children',
+      'content',
+      'lists',
+      'permissions',
+      'groups',
+      'members',
+      'owners',
+    ]);
+  }
 
   public async execute(context: Context): Promise<void> {
     if (!this.nextMiddleware) throw new Error('Next middleware not set');
 
     const endpoint = this.extractEndpoint(context.request);
-    const method = this.extractMethod(context.options);
+    const httpMethod = this.extractMethod(context.options);
+    const apiMethod = this.extractApiMethod(endpoint, httpMethod);
     const startTime = Date.now();
 
     try {
       await this.nextMiddleware.execute(context);
 
-      const duration = Date.now() - startTime;
-      const statusCode = context.response?.status || 0;
-      const statusClass = this.getStatusClass(statusCode);
+      const statusClass = getHttpStatusCodeClass(context.response?.status || 0);
+
+      this.spcGraphApiRequestDurationSeconds.record(elapsedSeconds(startTime), {
+        ms_tenant_id: this.msTenantId,
+        api_method: apiMethod,
+        result: 'success',
+        http_status: statusClass,
```

Comment:
```suggestion
        http_status_class: statusClass,
```
Just proposal since else its a bit mixed (I might have written it wrong in the ticket, maybe, but still).

_Applies to all others below won't further mention._

Reply:
Sure, sounds good :) Changed